python/ray/data/tests/test_backpressure_policies.py

```diff
@@ -291,6 +298,88 @@ def test_no_deadlock(self):
         )
 
 
+def test_large_e2e_backpressure(shutdown_only, restore_data_context):  # noqa: F811
+    """Test backpressure on a synthetic large-scale workload."""
+    # The cluster has 16 CPUs and 2GB object store memory.
+    # The dataset will have 2GB * 25% = 512MB memory budget.
+    #
+    # Each produce task generates 10 blocks, each of which has 0.1GB data.
+    #
+    # Without any backpressure, the producer tasks will output at most
+    # 16 * 10 * 0.1GB = 16GB data.
+    #
+    # With StreamingOutputBackpressurePolicy and the following configuration,
+    # the executor will still schedule 16 produce tasks, but only the first task is
+    # allowed to output all blocks. The total size of pending blocks will be
+    # (10 + 15 * 1 + 1) * 0.1GB = 2.6GB,
+    # where
+    # - 10 is the number of blocks in the first task.
+    # - 15 * 1 is the number of blocks pending at the streaming generator level of
+    #   the other 15 tasks.
+    # - 1 is the number of blocks pending at the output queue.
+    max_pending_block_bytes = int(2.6 * 1024**3)
+
+    ray.init(num_cpus=16, object_store_memory=2 * 1024 * 1024 * 1024)
+
+    NUM_ROWS_PER_TASK = 10
+    NUM_TASKS = 32
+    NUM_ROWS_TOTAL = NUM_ROWS_PER_TASK * NUM_TASKS
+    BLOCK_SIZE = 100 * 1024 * 1024
+
+    def produce(batch):
+        print("Produce task started", batch["id"])
+        time.sleep(0.1)
+        for id in batch["id"]:
+            print("Producing", id)
+            yield {
+                "id": [id],
+                "image": [np.zeros(BLOCK_SIZE, dtype=np.uint8)],
+            }
+
+    def consume(batch):
+        print("Consume task started", batch["id"])
+        time.sleep(0.1)
+        return {"id": batch["id"], "result": [0 for _ in batch["id"]]}
+
+    data_context = ray.data.DataContext.get_current()
+    data_context.set_config(
+        ENABLED_BACKPRESSURE_POLICIES_CONFIG_KEY,
+        [
+            StreamingOutputBackpressurePolicy,
+        ],
+    )
+    data_context.set_config(
+        StreamingOutputBackpressurePolicy.MAX_BLOCKS_IN_OP_OUTPUT_QUEUE_CONFIG_KEY, 1
+    )
+    data_context.set_config(
+        StreamingOutputBackpressurePolicy.MAX_BLOCKS_IN_GENERATOR_BUFFER_CONFIG_KEY, 1
+    )
+    data_context.execution_options.verbose_progress = True
+    data_context.target_max_block_size = BLOCK_SIZE
+
+    last_snapshot = get_initial_core_execution_metrics_snapshot()
+
+    ds = ray.data.range(NUM_ROWS_TOTAL, parallelism=NUM_TASKS)
+    ds = ds.map_batches(produce, batch_size=NUM_ROWS_PER_TASK)
+    ds = ds.map_batches(consume, batch_size=None, num_cpus=0.9)
+    for _ in ds.iter_batches():
+        pass
+    max_spilled_bytes = int(2.0 * max_pending_block_bytes)
+    last_snapshot = assert_core_execution_metrics_equals(
```

Comment:
It would be good to also add checks during the execution for cumulative_created_plasma_bytes. That way we can make sure that pauses in the iteration loop properly trigger backpressure.

Reply:
I tried adding an `assert_core_execution_metrics_equals` in the `iter_batches` loop, but that made the test too slow. Currently we assert the total spilled data amount at end of execution, I guess this can already make sure the tasks are paused. Do you have better ideas?

Reply:
Hmm maybe just do it once instead of on every loop? Or use a larger batch_size.